{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import random as rand\n",
    "import os\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "nucleotids = ['G', 'C', 'A', 'T']\n",
    "complementary = {\n",
    "  \"C\": \"G\",\n",
    "  \"G\": \"C\",\n",
    "  \"T\": \"A\", \n",
    "  \"A\": \"T\"\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Check if simulator input parameters are valid\n",
    "def checkArguments(avgQuality, coverage, readLength, insertSize, errRateSNV, errRateInsert, errRateDelete):\n",
    "    if(avgQuality < 0 or avgQuality > 40):\n",
    "        print(\"Average quaility value must be between 0 and 40!\")\n",
    "        return False\n",
    "    if(readLength*2 > insertSize):\n",
    "        print(\"Insert size must be at least two times read length!\")\n",
    "        return False\n",
    "    if(errRateInsert < 0 or errRateInsert > 1):\n",
    "        print(\"Insetion error rate is probability value and must be between 0 and 1!\")\n",
    "        return False\n",
    "    if(errRateDelete < 0 or errRateDelete > 1):\n",
    "        print(\"Deletion error rate is probability value and must be between 0 and 1!\")\n",
    "        return False\n",
    "    if(errRateSNV < 0 or errRateSNV > 1):\n",
    "        print(\"SNV error rate is probability value and must be between 0 and 1!\")\n",
    "        return False\n",
    "    return True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Read all sequences from fasta file\n",
    "def parseFastaFile(file):\n",
    "    name, seq = None, []\n",
    "    for line in file:\n",
    "        line = line.rstrip()\n",
    "        if line.startswith(\">\"):\n",
    "            if name: yield (name, ''.join(seq))\n",
    "            name, seq = line, []\n",
    "        else:\n",
    "            seq.append(line)\n",
    "    if name: yield (name, ''.join(seq))\n",
    "\n",
    "#Parse reference genome\n",
    "def parseGenomeFile(fileName):\n",
    "    refGenome = {}\n",
    "    with open(fileName) as file:\n",
    "        for name, seq in parseFastaFile(file):\n",
    "            refGenome[name[1:]] = list(seq)\n",
    "    return refGenome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "DeletionMutations = {}\n",
    "InsertionMutations = {}\n",
    "SNVMutations = {}\n",
    "\n",
    "def insertMutations(refGenome, errRateSNV, errRateInsert, errRateDelete):\n",
    "    totalMutations = 0\n",
    "    for name,sequence in refGenome.items():            \n",
    "        SNVnum = round(len(sequence)*errRateSNV)\n",
    "        DeleteNum = round(len(sequence)*errRateDelete)\n",
    "        InsertNum = round(len(sequence)*errRateInsert)\n",
    "        \n",
    "        totalMutations += SNVnum\n",
    "        totalMutations += DeleteNum\n",
    "        totalMutations += InsertNum\n",
    "        \n",
    "        delList = []\n",
    "        insertList = []\n",
    "        snvList = []\n",
    "        \n",
    "            \n",
    "        #Deletion mutations\n",
    "        for i in range(DeleteNum):\n",
    "            mutationPosition = rand.randint(0,len(sequence)-1)\n",
    "            del sequence[mutationPosition]\n",
    "            delList.append(mutationPosition)\n",
    "        \n",
    "        #Insertion mutations\n",
    "        for i in range(InsertNum):\n",
    "            mutationPosition = rand.randint(0,len(sequence)-1)\n",
    "            insertionNucleotid = rand.randint(0,3)\n",
    "            sequence.insert(mutationPosition,nucleotids[insertionNucleotid])\n",
    "            insertList.append(mutationPosition)\n",
    "            for deletion in delList:\n",
    "                if(deletion >= mutationPosition):\n",
    "                    deletion += 1\n",
    "            \n",
    "        #SNV mutations\n",
    "        for i in range(SNVnum):\n",
    "            mutationPosition = rand.randint(0,len(sequence)-1)\n",
    "            mutationNucleotid = rand.randint(0,3)\n",
    "            while(sequence[mutationPosition] == nucleotids[mutationNucleotid]):\n",
    "                mutationNucleotid = rand.randint(0,3)\n",
    "            sequence[mutationPosition] = nucleotids[mutationNucleotid]\n",
    "            snvList.append(mutationPosition)\n",
    "            \n",
    "        DeletionMutations[name] = delList\n",
    "        InsertionMutations[name] = insertList\n",
    "        SNVMutations[name] = snvList\n",
    "    \n",
    "    return totalMutations\n",
    "                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "#generate nucleotids qualities using normal distribution\n",
    "def generateQuality(avgQuality, readLength):\n",
    "    quality = []\n",
    "    randomNormal = np.random.normal(avgQuality,5,readLength)\n",
    "    for i in range(len(randomNormal)):\n",
    "        if(randomNormal[i] < 0):\n",
    "            randomNormal[i] = 0\n",
    "        if(randomNormal[i] > 40):\n",
    "            randomNormal[i] = 40\n",
    "        quality.append(chr(int(round(randomNormal[i])) + 33))\n",
    "    return \"\".join(quality)\n",
    "\n",
    "#calculate position of the base in reference genome\n",
    "def getOriginalPosition(position,name):\n",
    "    insertCnt = 0\n",
    "    deleteCnt = 0\n",
    "    for insertion in InsertionMutations[name]:\n",
    "        if(insertion < position):\n",
    "            insertCnt += 1\n",
    "    for deletion in DeletionMutations[name]:\n",
    "        if(deletion < position):\n",
    "            deleteCnt += 1\n",
    "    return position - insertCnt + deleteCnt\n",
    "\n",
    "def sequenceReads(refGenomeFile,refGenome,avgQuality,coverage,readLength,insertSize):\n",
    "    #get file name\n",
    "    fileNameWithExtension = os.path.basename(refGenomeFile)\n",
    "    fileName = os.path.splitext(fileNameWithExtension)[0]\n",
    "    #create FASTQ and SAM files\n",
    "    samFile = open(fileName + \"_sim.sam\",\"w\")\n",
    "    firstReadFile = open(fileName + \"_1.fastq\",\"w\")\n",
    "    secondReadFile = open(fileName + \"_2.fastq\",\"w\")\n",
    "    \n",
    "    totalReads = 0\n",
    "    \n",
    "    for name,sequence in refGenome.items():\n",
    "        pairedReadsNum = round((coverage*len(sequence))/(2*readLength))\n",
    "        totalReads += pairedReadsNum\n",    
    "        covered = np.zeros(len(sequence))\n",
    "        for i in range(pairedReadsNum):\n",
    "            #choose insert position so it is not already covered\n",
    "            insertPosition = rand.randint(0,len(sequence)-insertSize)\n",
    "            while(covered[insertPosition] == 1):\n",    
    "                insertPosition = rand.randint(0,len(sequence)-insertSize)\n",
    "            firstReadEnd = insertPosition + readLength\n",
    "            secondReadStart = insertPosition + insertSize - readLength\n",
    "            secondReadEnd = insertPosition + insertSize\n",
    "            covered[insertPosition:firstReadEnd] = [1] * readLength\n",
    "            covered[secondReadStart:secondReadEnd] = [1] * readLength\n",
    "            #adjust insert position so neither of reads start with mutated base\n",
    "            while((insertPosition + insertSize < len(sequence))):\n",
    "                if((insertPosition in SNVMutations[name]) or ((insertPosition + insertSize - readLength) in SNVMutations[name])):\n",
    "                    insertPosition += 1\n",
    "                elif ((insertPosition in InsertionMutations[name]) or ((insertPosition + insertSize - readLength) in InsertionMutations[name])):\n",
    "                    insertPosition += 1\n",
    "                else:\n",
    "                    break\n",
    "            \n",
    "            seqId = \"read_\" + str(i) + \":\" + str(insertPosition + 1) + \":\" + str(insertPosition + insertSize + 1)\n",
    "            \n",
    "            #sequence first read and write it to files\n",
    "            firstRead = []\n",
    "            for j in range(readLength):\n",
    "                firstRead.append(sequence[insertPosition + j])\n",
    "            quality = generateQuality(avgQuality, readLength)\n",
    "            firstReadFile.write(\"@\" + str(seqId) + \"/1\\n\" + \"\".join(firstRead) + \"\\n+\\n\" + quality + \"\\n\")\n",
    "            tab = \"\\t\"\n",
    "            pos = getOriginalPosition(insertPosition,name)\n",
    "            samFile.write(seqId+tab+str(pos+1)+tab+\"\".join(firstRead)+tab+quality+\"\\n\")\n",
    "            \n",
    "            #sequence second read and write it to files (it is inverse and complementary)\n",
    "            secondRead = []\n",
    "            originalSecond = []\n",
    "            for j in range(readLength):\n",
    "                if(sequence[insertPosition + insertSize - j - 1] in nucleotids):\n",
    "                    secondRead.append(complementary[sequence[insertPosition + insertSize - j - 1]])\n",
    "                else:\n",
    "                    secondRead.append(sequence[insertPosition + insertSize - j - 1])\n",
    "                originalSecond.append(sequence[insertPosition + insertSize - readLength + j])\n",
    "            quality = generateQuality(avgQuality, readLength)\n",
    "            secondReadFile.write(\"@\" + str(seqId) + \"/2\\n\" + \"\".join(secondRead) + \"\\n+\\n\" + quality + \"\\n\")\n",
    "            pos = getOriginalPosition(insertPosition+insertSize-readLength,name)\n",
    "            samFile.write(seqId+tab+str(pos+1)+tab+\"\".join(originalSecond)+tab+quality+\"\\n\")\n",
    "            \n",
    "    samFile.close()\n",
    "    firstReadFile.close()\n",
    "    secondReadFile.close()\n",
    "    return totalReads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "def simulate(refGenomeFile, avgQuality, coverage, readLength, insertSize, errRateSNV, errRateInsert, errRateDelete):\n",
    "    if(checkArguments(avgQuality, coverage, readLength, insertSize, errRateSNV, errRateInsert, errRateDelete)):\n",
    "        refGenome = parseGenomeFile(refGenomeFile)\n",
    "        insertMutations(refGenome,errRateSNV,errRateInsert,errRateDelete)\n",
    "        sequenceReads(refGenomeFile,refGenome,avgQuality,coverage,readLength,insertSize)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def evaluateBWAMEM(fileName):\n",
    "    simulatedSam = open(fileName+\"_sim.sam\", \"r\")\n",
    "    bwamemSam = open(fileName+\"_bwa.sam\", \"r\")\n",
    "    alignments = {}\n",
    "    aligned = 0\n",
    "    for read in simulatedSam.readlines():\n",
    "        split = read.split(\"\\t\")\n",
    "        alignments[split[0]] = split[1]\n",
    "    for read in bwamemSam.readlines():\n",
    "        if(read[0] == \"@\"):\n",
    "            continue\n",
    "        split = read.split(\"\\t\")\n",
    "        if(split[3] == alignments[split[0]]):\n",
    "            aligned += 1\n",
    "    simulatedSam.close()\n",
    "    bwamemSam.close()\n",
    "    print(\"Correctly aligned \"+str(aligned)+\" of \"+str(len(alignments))+\" reads, rate \"+str(aligned/len(alignments)*100)+\"%\")\n",
    "        "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
